drivers/serial/uart_numaker.c

```diff
@@ -0,0 +1,451 @@
+
+	return 0;
+}
+#endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
+
+static int uart_numaker_init(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	struct uart_numaker_data *pData = dev->data;
+	int err = 0;
+
+	SYS_UnlockReg();
+
+	struct numaker_scc_subsys scc_subsys;
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = config->clk_modidx;
+	scc_subsys.pcc.clk_src = config->clk_src;
+	scc_subsys.pcc.clk_div = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock(clk_modidx) */
+	err = clock_control_on(config->clk_dev, (clock_control_subsys_t)&scc_subsys);
+	if (err != 0) {
+		goto move_exit;
+	}
+	/* Equivalent to CLK_SetModuleClock(clk_modidx, clk_src, clk_div) */
+	err = clock_control_configure(config->clk_dev, (clock_control_subsys_t)&scc_subsys, NULL);
+	if (err != 0) {
+		goto move_exit;
+	}
+
+	/*
+	 * Set pinctrl for UART0 RXD and TXD
+	 * Set multi-function pins for UART0 RXD and TXD
+	 */
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto move_exit;
+	}
+
+	/* Same as BSP's SYS_ResetModule(id_rst) */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	/* Reset UART to default state */
+	reset_line_toggle_dt(&config->reset);
+
+	UART_Open(config->uart, pData->ucfg.baudrate);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	config->irq_config_func(dev);
+#endif
+
+move_exit:
+	SYS_LockReg();
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int uart_numaker_fifo_fill(const struct device *dev, const uint8_t *tx_data, int size)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+	int tx_bytes = 0;
+
+	/* Check TX FIFO not full, then fill */
+	while (((size - tx_bytes) > 0) && (!(uart->FIFOSTS & UART_FIFOSTS_TXFULL_Msk))) {
+		/* Fill one byte into TX FIFO */
+		uart->DAT = tx_data[tx_bytes++];
+	}
+
+	return tx_bytes;
+}
+
+static int uart_numaker_fifo_read(const struct device *dev, uint8_t *rx_data, const int size)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+	int rx_bytes = 0;
+
+	/* Check RX FIFO not empty, then read */
+	while (((size - rx_bytes) > 0) && (!(uart->FIFOSTS & UART_FIFOSTS_RXEMPTY_Msk))) {
+		/* Read one byte from UART RX FIFO */
+		rx_data[rx_bytes++] = (uint8_t)uart->DAT;
+	}
+
+	return rx_bytes;
+}
+
+static void uart_numaker_irq_tx_enable(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+
+	UART_EnableInt(uart, UART_INTEN_THREIEN_Msk);
+}
+
+static void uart_numaker_irq_tx_disable(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+
+	UART_DisableInt(uart, UART_INTEN_THREIEN_Msk);
+}
+
+#define UART_ENABLE_INT(uart, u32eIntSel) ((uart)->INTEN |= (u32eIntSel))
```

Comment:
Remove?

Reply:
@dcpleung,
Thanks, it's redundancy and I removed it in commit [01d6eec](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/01d6eecb045d76b6ed5c0911a9d8c29ea0d9f0a2).